src/icosphere.rs

```diff
@@ -0,0 +1,207 @@
+//   Copyright GFX Developers 2014-2017
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+//! Icosahedral sphere
+
+use std::collections::HashMap;
+
+use cgmath::{InnerSpace, Vector3};
+
+use {Vertex, Triangle};
+use generators::{IndexedPolygon, SharedVertex};
+
+/// Icosahedral sphere with radius 1, centered at (0., 0., 0.)
+#[derive(Clone, Debug)]
+pub struct IcoSphere {
+    i: usize,
+    vertices: Vec<[f32; 3]>,
+    faces: Vec<[usize; 3]>,
+}
+
+const T: f32 = 0.85065080835204;
```

Comment:
where do these come from?

Reply:
They come from calculating the sides of a rectangle inside the sphere.

Reply:
An icosahedrons vertices can be defined as the corner points of three orthogonal golden rectangles, which means the sides of the rectangles have the golden ratio: (1 + sqrt(5)) / 2. If you take that directly though you won't get a unit sphere, so we need to normalize the points (0, ±1, ±ratio), which gives those 2 values. I precomputed those so it wouldn't be necessary runtime.